pubsub/producer.go

```diff
@@ -0,0 +1,242 @@
+	"github.com/spf13/pflag"
+)
+
+const (
+	messageKey   = "msg"
+	defaultGroup = "default_consumer_group"
+)
+
+type Marshallable[T any] interface {
+	Marshal() []byte
+	Unmarshal(val []byte) (T, error)
+}
+
+type Producer[Request Marshallable[Request], Response Marshallable[Response]] struct {
+	stopwaiter.StopWaiter
+	id     string
+	client redis.UniversalClient
+	cfg    *ProducerConfig
+
+	promisesLock sync.RWMutex
+	promises     map[string]*containers.Promise[Response]
+}
+
+type ProducerConfig struct {
+	RedisURL string `koanf:"redis-url"`
+	// Redis stream name.
+	RedisStream string `koanf:"redis-stream"`
+	// Interval duration in which producer checks for pending messages delivered
+	// to the consumers that are currently inactive.
+	CheckPendingInterval time.Duration `koanf:"check-pending-interval"`
+	// Duration after which consumer is considered to be dead if heartbeat
+	// is not updated.
+	KeepAliveTimeout time.Duration `koanf:"keepalive-timeout"`
+	// Interval duration for checking the result set by consumers.
+	CheckResultInterval time.Duration `koanf:"check-result-interval"`
+	// Redis consumer group name.
+	RedisGroup string `koanf:"redis-group"`
+}
+
+var DefaultProducerConfig = &ProducerConfig{
+	RedisStream:          "default",
+	CheckPendingInterval: time.Second,
+	KeepAliveTimeout:     5 * time.Minute,
+	CheckResultInterval:  5 * time.Second,
+	RedisGroup:           defaultGroup,
+}
+
+var DefaultTestProducerConfig = &ProducerConfig{
+	RedisStream:          "default",
+	RedisGroup:           defaultGroup,
+	CheckPendingInterval: 10 * time.Millisecond,
+	KeepAliveTimeout:     20 * time.Millisecond,
+	CheckResultInterval:  5 * time.Millisecond,
+}
+
+func ProducerAddConfigAddOptions(prefix string, f *pflag.FlagSet) {
+	f.String(prefix+".redis-url", DefaultConsumerConfig.RedisURL, "redis url for redis stream")
+	f.Duration(prefix+".response-entry-timeout", DefaultConsumerConfig.ResponseEntryTimeout, "timeout for response entry")
+	f.Duration(prefix+".keepalive-timeout", DefaultConsumerConfig.KeepAliveTimeout, "timeout after which consumer is considered inactive if heartbeat wasn't performed")
+	f.String(prefix+".redis-stream", DefaultConsumerConfig.RedisStream, "redis stream name to read from")
+	f.String(prefix+".redis-group", DefaultConsumerConfig.RedisGroup, "redis stream consumer group name")
+}
+
+func NewProducer[Request Marshallable[Request], Response Marshallable[Response]](cfg *ProducerConfig) (*Producer[Request, Response], error) {
+	if cfg.RedisURL == "" {
+		return nil, fmt.Errorf("redis url cannot be empty")
+	}
+	c, err := redisutil.RedisClientFromURL(cfg.RedisURL)
+	if err != nil {
+		return nil, err
+	}
+	return &Producer[Request, Response]{
+		id:       uuid.NewString(),
+		client:   c,
+		cfg:      cfg,
+		promises: make(map[string]*containers.Promise[Response]),
+	}, nil
+}
+
+func (p *Producer[Request, Response]) Start(ctx context.Context) {
+	p.StopWaiter.Start(ctx, p)
+	p.StopWaiter.CallIteratively(
```

Comment:
Not a must - but could we move those two callIteratively to the first time someone calls Produce?
This iterative approach has interesting advantages over each "produce" spawning it's own thread. The one disadvantage seems to me to be that even on systems where the entire interface is unused.